### ISSUE TITLE
 Error compiling Cython file during [make] progress:     cysteps_dist.pyx:32:8: 'mpi4py/MPI.pxd' not found

### ISSUE THREAD

**Issue by FireworkZhao** [CNS-OIST/STEPS#40]
When compling the latest version,  the source code can't be complied due to the following problem even through prerequisites have been installed.

[ 62%] Cythonizing extension cysteps_dist.pyx

Error compiling Cython file:
------------------------------------------------------------
...
This file is the user-interface file for all objects related to STEPS 4. All
objects are directly derived from the corresponding Cython objects.
"""

from cython.operator cimport dereference as deref
cimport mpi4py.MPI as MPI
       ^
------------------------------------------------------------

cysteps_dist.pyx:32:8: 'mpi4py/MPI.pxd' not found
make[2]: *** [pysteps/CMakeFiles/cysteps.dir/build.make:98: ../pysteps/steps/cysteps_dist.cpp] Error 1
make[2]: *** Deleting file '../pysteps/steps/cysteps_dist.cpp'
make[1]: *** [CMakeFiles/Makefile2:1614: pysteps/CMakeFiles/cysteps.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

**Comment by WeiliangChenOIST** [comment]
continue from #38 and #39

**Comment by WeiliangChenOIST** [comment]
Dear Zhao,
Could you run the following steps and copy the terminal outputs here so that we have more information please?
assuming that you are under the STEPS directory you've cloned
```
rm -rf build
mkdir build
cd build
cmake ..
```
You may need to add extra parameters for the cmake command if you did so before.

**Comment by FireworkZhao** [comment]
> Dear Zhao, Could you run the following steps and copy the terminal outputs here so that we have more information please? assuming that you are under the STEPS directory you've cloned
> 
> ```
> rm -rf build
> mkdir build
> cd build
> cmake ..
> ```
> 
> You may need to add extra parameters for the cmake command if you did so before.

Yes，I have done that and the result shows below:
-- The CXX compiler identification is GNU 11.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- STEPS_TEST_FORMATTING: OFF
-- STEPS_GIT_HOOKS: OFF
-- STEPS_GIT_COMMIT_HOOKS: 
-- STEPS_GIT_PUSH_HOOKS: courtesy-msg
-- STEPS_STATIC_ANALYSIS: OFF
-- STEPS_TEST_STATIC_ANALYSIS: OFF
-- Added modern target Omega_h::omega_h
-- Added modern target SUNDIALS::SUNDIALS
-- Found METIS: /usr/lib/x86_64-linux-gnu/libmetis.so  
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Performing Test COMPILER_SUPPORTS_CXX14
-- Performing Test COMPILER_SUPPORTS_CXX14 - Success
-- Performing Test COMPILER_SUPPORTS_CXX17
-- Performing Test COMPILER_SUPPORTS_CXX17 - Success
-- Generating debug symbols: -ggdb3
-- Looking for strip utility
-- Looking for strip utility - found
-- Found Python: /home/STEPS/venv/bin/python3.10 (found suitable version "3.10.12", minimum required is "3.6") found components: Interpreter Development Development.Module Development.Embed 
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Found GSL: /usr/include (found version "2.7.1") 
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0")  
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- Looking for sgemm_
-- Looking for sgemm_ - not found
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so  
-- Looking for cheev_
-- Looking for cheev_ - found
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl  
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Modern target BLAS::BLAS already exists, skipping setup
-- Modern target LAPACK::LAPACK already exists, skipping setup
-- Found build: /home/STEPS/venv/lib/python3.10/site-packages/build (found version "1.2.1") 
-- Found piptools: /home/STEPS/venv/lib/python3.10/site-packages/piptools  
-- Found PythonInterp: /home/STEPS/venv/bin/python (found version "3.10.12") 
-- Found Cython: /home/STEPS/venv/bin/cython (found suitable version "3.0.10", minimum required is "0.29") 
-- Found Gmsh: /usr/lib/x86_64-linux-gnu/libgmsh.so (found version "4.8.4") 
-- Caliper: OFF
-- Likwid: OFF
-- Region tracker: OFF
-- Performing Test HAVE_FLAG__ffile_prefix_map__home_steps_STEPS_build__deps_catch2_src__
-- Performing Test HAVE_FLAG__ffile_prefix_map__home_steps_STEPS_build__deps_catch2_src__ - Success
-- STEPS PYTHON packages will be installed to /home/STEPS/venv/lib/python3.10/site-packages
To change use -DPython_INSTALL_PREFIX=<dir>
-- Bundled libraries will be installed to /home/zyh/.local/lib
To change use -DCMAKE_INSTALL_PREFIX=<dir> (without the lib directory)
-- Found steps: /home/STEPS/venv/lib/python3.10/site-packages/steps-3.5.0-py3.10.egg/steps (found version "3.5.0") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/steps/STEPS/build

Usually, the problem occurred after the step [make]. Thank you for your time.

**Comment by WeiliangChenOIST** [comment]
thanks. What about the output of this:

```
python -c "import mpi4py;print(mpi4py.get_include())"
```

**Comment by FireworkZhao** [comment]
> python -c "import mpi4py;print(mpi4py.get_include())"

The answer would be: 
(venv)   :~/steps/STEPS/build$ python -c "import mpi4py;print(mpi4py.get_include())"
/home/STEPS/venv/lib/python3.10/site-packages/mpi4py/include

I wonder if I need to completely delete STEPS 3.5.0 then build it instead of trying to upgrade it from 3.5.0 to the latest version.

**Comment by WeiliangChenOIST** [comment]
> I wonder if I need to completely delete STEPS 3.5.0 then build it instead of trying to upgrade it from 3.5.0 to the latest version.
That shouldn't matter.

Could you go to `/home/STEPS/venv/lib/python3.10/site-packages/mpi4py/` and see if you can find `MPI.pxd` in the directory please?

**Comment by WeiliangChenOIST** [comment]
@tristan0x is it possible to reproduce this on a Linux machine, or do you have any clue what may cause it?

**Comment by FireworkZhao** [comment]
> > I wonder if I need to completely delete STEPS 3.5.0 then build it instead of trying to upgrade it from 3.5.0 to the latest version.
> > That shouldn't matter.

Despite of this, I tried to remove STEPS 3.5.0 and recompiled the whole project.
I have done the following steps:
cmake ..
make
sudo make install

Then I get the result:
Installing collected packages: steps
Successfully installed steps-5.0.1
-- Installing: /home/STEPS/venv/lib/python3.10/site-packages/steps/cysteps_dist.so
-- Set runtime path of "/home/STEPS/venv/lib/python3.10/site-packages/steps/cysteps_dist.so" to "/home/zyh/.local/lib:/usr/lib/x86_64-linux-gnu/openmpi/lib"
-- Up-to-date: /home/.local/lib/libomega_h.so
-- Up-to-date: /home/.local/lib
-- Up-to-date: /home/.local/lib/libsundials_nvecserial.so.3.2.1
-- Up-to-date: /home/.local/lib/libsundials_nvecserial.so.3
-- Up-to-date: /home/.local/lib/libsundials_nvecserial.so
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixdense.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixdense.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixdense.so
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixband.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixband.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixband.so
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixsparse.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixsparse.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunmatrixsparse.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolband.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolband.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolband.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsoldense.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsoldense.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsoldense.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspgmr.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspgmr.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspgmr.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspfgmr.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspfgmr.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspfgmr.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspbcgs.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspbcgs.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolspbcgs.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolsptfqmr.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolsptfqmr.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolsptfqmr.so
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolpcg.so.1.2.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolpcg.so.1
-- Up-to-date: /home/.local/lib/libsundials_sunlinsolpcg.so
-- Up-to-date: /home/.local/lib/libsundials_arkode.so.2.2.1
-- Up-to-date: /home/.local/lib/libsundials_arkode.so.2
-- Up-to-date: /home/.local/lib/libsundials_arkode.so
-- Up-to-date: /home/.local/lib/libsundials_cvode.so.3.2.1
-- Up-to-date: /home/.local/lib/libsundials_cvode.so.3
-- Up-to-date: /home/.local/lib/libsundials_cvode.so
-- Up-to-date: /home/.local/lib/libsundials_ida.so.3.2.1
-- Up-to-date: /home/.local/lib/libsundials_ida.so.3
-- Up-to-date: /home/.local/lib/libsundials_ida.so
-- Up-to-date: /home/.local/lib/libsundials_kinsol.so.3.2.1
-- Up-to-date: /home/.local/lib/libsundials_kinsol.so.3
-- Up-to-date: /home/.local/lib/libsundials_kinsol.so

But the result of testing intallation by running the command [python -c "import steps; steps._greet()"] shows:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/STEPS/venv/lib/python3.10/site-packages/steps/__init__.py", line 48, in <module>
    from . import cysteps_dist
ImportError: /home/STEPS/venv/lib/python3.10/site-packages/steps/cysteps_dist.so: undefined symbol: _ZN5steps4dist12OhmicCurrent5resetEv

As I put [pip list] in terminal, I do can see [setps 5.0.1] listed below.

In conclusion, I wonder are there any errors in building  [cysteps_dist.so], if so how can I fix it?

**Comment by WeiliangChenOIST** [comment]
thanks for the update. We can reproduce this error on our system now, will look into the cause of it.

**Comment by jlallouette** [comment]
Dear @FireworkZhao,

Did you install PETSc on your system? If not, can you try to install it and re-run the `cmake ..`, `make` and `make install` commands?
The issue you are experiencing is probably linked to PETSc not being installed and STEPS defaulting to installing the `DistTetOpSplit` solver that partially depends on PETSc. With the current release, you should be able to overcome it by either installing PETSc or by giving the `-DSTEPS_USE_DIST_MESH=False` option to the `cmake` command.

We are working on a fix to prevent a missing PETSc from generating the error you got.

**Comment by FireworkZhao** [comment]
> Dear @FireworkZhao,
> 
> Did you install PETSc on your system? If not, can you try to install it and re-run the `cmake ..`, `make` and `make install` commands? The issue you are experiencing is probably linked to PETSc not being installed and STEPS defaulting to installing the `DistTetOpSplit` solver that partially depends on PETSc. With the current release, you should be able to overcome it by either installing PETSc or by giving the `-DSTEPS_USE_DIST_MESH=False` option to the `cmake` command.
> 
> We are working on a fix to prevent a missing PETSc from generating the error you got.



> Dear @FireworkZhao,
> 
> Did you install PETSc on your system? If not, can you try to install it and re-run the `cmake ..`, `make` and `make install` commands? The issue you are experiencing is probably linked to PETSc not being installed and STEPS defaulting to installing the `DistTetOpSplit` solver that partially depends on PETSc. With the current release, you should be able to overcome it by either installing PETSc or by giving the `-DSTEPS_USE_DIST_MESH=False` option to the `cmake` command.
> 
> We are working on a fix to prevent a missing PETSc from generating the error you got.

thank you for your reply.
I have followed your instruction of installing PETSc and ru-run the whole 'cmake.. ; make; make install' steps, a new problem seems to occured:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/zyh/STEPS/venv/lib/python3.10/site-packages/steps/__init__.py", line 48, in <module>
    from . import cysteps_dist
  File "cysteps_dist.pyx", line 1, in init cysteps_dist
ValueError: mpi4py.MPI.Datatype size changed, may indicate binary incompatibility. Expected 40 from C header, got 32 from PyObject
I wonder if this problem still involves PETSc?

**Comment by jlallouette** [comment]
Dear @FireworkZhao,

It does seem to be related to an mpi4py version issue, can you confirm that you are using mpi4py 3.1.6?
Could you try to clear your `build` folder, remove the `/home/STEPS/venv/lib/python3.10/site-packages/steps` folder, run `cmake ..`, `make`, `make install` and attach the results of each of these commands?
It also seems that you are using a virtual environment, if so, make sure that it is activated (with `source /path/to/env/bin/activate` if you use `venv`) before the `cmake ..`, `make`, and `make install` commands.
If clearing the `build` folder doesn't work, can you try to run the `cmake -DSTEPS_USE_DIST_MESH=False ..` command instead of `cmake ..` to build STEPS without the distributed solver?

**Comment by FireworkZhao** [comment]
> Dear @FireworkZhao,
> 
> It does seem to be related to an mpi4py version issue, can you confirm that you are using mpi4py 3.1.6? Could you try to clear your `build` folder, remove the `/home/STEPS/venv/lib/python3.10/site-packages/steps` folder, run `cmake ..`, `make`, `make install` and attach the results of each of these commands? It also seems that you are using a virtual environment, if so, make sure that it is activated (with `source /path/to/env/bin/activate` if you use `venv`) before the `cmake ..`, `make`, and `make install` commands. If clearing the `build` folder doesn't work, can you try to run the `cmake -DSTEPS_USE_DIST_MESH=False ..` command instead of `cmake ..` to build STEPS without the distributed solver?

I've checked the mpi4py version which is indeed 3.1.6.
The result of 'cmake ..' is:
-- The CXX compiler identification is GNU 11.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- STEPS_TEST_FORMATTING: OFF
-- STEPS_GIT_HOOKS: OFF
-- STEPS_GIT_COMMIT_HOOKS: 
-- STEPS_GIT_PUSH_HOOKS: courtesy-msg
-- STEPS_STATIC_ANALYSIS: OFF
-- STEPS_TEST_STATIC_ANALYSIS: OFF
-- Added modern target Omega_h::omega_h
-- Added modern target SUNDIALS::SUNDIALS
-- Found METIS: /usr/lib/x86_64-linux-gnu/libmetis.so  
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Performing Test COMPILER_SUPPORTS_CXX14
-- Performing Test COMPILER_SUPPORTS_CXX14 - Success
-- Performing Test COMPILER_SUPPORTS_CXX17
-- Performing Test COMPILER_SUPPORTS_CXX17 - Success
-- Generating debug symbols: -ggdb3
-- Looking for strip utility
-- Looking for strip utility - found
-- Found Python: /home/STEPS/venv/bin/python3.10 (found suitable version "3.10.12", minimum required is "3.6") found components: Interpreter Development Development.Module Development.Embed 
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Found GSL: /usr/include (found version "2.7.1") 
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0")  
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- PETSc found: /usr/lib/petscdir/petsc3.15/x86_64-linux-gnu-real/lib
-- Looking for sgemm_
-- Looking for sgemm_ - not found
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so  
-- Looking for cheev_
-- Looking for cheev_ - found
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/libopenblas.so;-lm;-ldl  
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Modern target BLAS::BLAS already exists, skipping setup
-- Modern target LAPACK::LAPACK already exists, skipping setup
-- Added modern target PETSC::PETSC
-- Found build: /home/STEPS/venv/lib/python3.10/site-packages/build (found version "1.2.1") 
-- Found piptools: /home/STEPS/venv/lib/python3.10/site-packages/piptools  
-- Found PythonInterp: /home/STEPS/venv/bin/python (found version "3.10.12") 
-- Found Cython: /home/STEPS/venv/bin/cython (found suitable version "3.0.10", minimum required is "0.29") 
-- Found Gmsh: /usr/lib/x86_64-linux-gnu/libgmsh.so (found version "4.8.4") 
-- Caliper: OFF
-- Likwid: OFF
-- Region tracker: OFF
-- Performing Test HAVE_FLAG__ffile_prefix_map__home_steps_STEPS_build__deps_catch2_src__
-- Performing Test HAVE_FLAG__ffile_prefix_map__home_steps_STEPS_build__deps_catch2_src__ - Success
-- STEPS PYTHON packages will be installed to /home/STEPS/venv/lib/python3.10/site-packages
To change use -DPython_INSTALL_PREFIX=<dir>
-- Bundled libraries will be installed to /home/.local/lib
To change use -DCMAKE_INSTALL_PREFIX=<dir> (without the lib directory)
-- Could NOT find steps (missing: STEPS_LOCATION _STEPS_VERSION_MATCH) 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/steps/STEPS/build

The reslut of 'make' is pasted in the attach file:
[result of [make].txt](https://github.com/CNS-OIST/STEPS/files/15387499/result.of.make.txt)

The result of 'make install' is pasted in the attach file:
[reslut of[make install].txt](https://github.com/CNS-OIST/STEPS/files/15388015/reslut.of.make.install.txt)

I have make sure that the virtual environment [venv] is activated.

After I run the command: python3 -c "import steps; steps._greet()"  and get the result:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'steps' has no attribute '_greet'

**Comment by iahepburn** [comment]
what does 
`python3 -c "import steps; print(steps.__version__)"
`
show?

**Comment by FireworkZhao** [comment]
> python3 -c "import steps; print(steps.__version__)"
It says:
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'steps' has no attribute '__version__'

**Comment by iahepburn** [comment]
did you give the underscores?

**Comment by FireworkZhao** [comment]
> did you give the underscores?

sorry for not replying in time.
Yes, I have.
![pic_result](https://github.com/CNS-OIST/STEPS/assets/91300417/0dad4b0c-f68b-44c7-beb6-b44cb3b15dcc)

**Comment by jlallouette** [comment]
Do you have a file called `steps.py` or a folder and file called `steps/__init__.py` in the directory in which you are running the command? If so, python is probably importing these instead of the steps package that you installed. Try to run the same commands but in a different folder. If this is not the problem, can you check that steps is installed in `/home/STEPS/venv/lib/python3.10/site-packages/steps`?

**Comment by FireworkZhao** [comment]
> Do you have a file called `steps.py` or a folder and file called `steps/__init__.py` in the directory in which you are running the command? If so, python is probably importing these instead of the steps package that you installed. Try to run the same commands but in a different folder. If this is not the problem, can you check that steps is installed in `/home/STEPS/venv/lib/python3.10/site-packages/steps`?

Thank you for your suggestion，I changed the folder and ran the command. It did solve the problem. Now I've got the result:
STochastic Engine for Pathway Simulation
Version:  5.0.1
License:  GPL3.0
Website:  steps.sourceforge.net
CXX Binding: Cython
and "5.0.1" for both.
Thanks again for your time and help!

**Comment by WeiliangChenOIST** [comment]
Resolved.
